### PR TITLE
Upgrade to Spring Boot 3.2

### LINKED ISSUE #24091: Upgrade to Spring Boot 3.2
##### **Overview of the issue**

According to [calendar.spring.io](https://calendar.spring.io/), Spring Boot 3.2 will be released on November 23rd, 2023. We should do a release with 3.2 support before the end of the year. I also think we should do some sort of tribute to the frameworks we're dependent on in a blog post because we're successful because of them. That'd be awesome for Thanksgiving. 

##### **Motivation for or Use Case**

If you'd like to be a part of https://dev.to/jhipster, please let us know!

### PR DESCRIPTION
**Pull request by mraible**
Verifies that https://github.com/jhipster/jhipster-bom/pull/1460 works. 
Fixes #24091.

---

Please make sure the below checklist is followed for Pull Requests.

- [x] [All continuous integration tests](https://github.com/jhipster/generator-jhipster/actions) are green
- [x] Tests are added where necessary
- [x] The JDL part is updated if necessary
- [x] [jhipster-online](https://github.com/jhipster/jhipster-online) is updated if necessary
- [x] Documentation is added/updated where necessary
- [x] Coding Rules & Commit Guidelines as per our [CONTRIBUTING.md document](https://github.com/jhipster/generator-jhipster/blob/main/CONTRIBUTING.md) are followed

### CODE REVIEW

**Comment by mraible** [comment]
Closing and re-opening to test updates to jhipster-bom.

**Comment by mraible** [comment]
Trying again because jhipster-bom has fixes. Please let me know if there's an easier way than closing/reopening.

**Comment by mraible** [comment]
It seems kinda strange that we have to add the `name` argument for `@PathVariable` and `@RequestParam` for Spring Boot 3.2. See https://github.com/jhipster/generator-jhipster/pull/24098/commits/56cebf91073faf9b855bbc8854794c148e4c9137. I found the answer [here](https://stackoverflow.com/a/25797744). 

/cc @mshima and @hide212131 since I know we had to do this for the generator-jhipster-native blueprint. It might not be necessary in the blueprint once this PR is merged.

**Comment by mraible** [comment]
I'm not able to reproduce the errors in the [ng-webflux-psql-additional](https://github.com/jhipster/generator-jhipster/actions/runs/6938688102/job/18874970827?pr=24098#logs) job locally. 

```
java.lang.NoClassDefFoundError: com/fasterxml/jackson/databind/ser/std/StdArraySerializers$DoubleArraySerializer
```

**Comment by mshima** [comment]
> It seems kinda strange that we have to add the `name` argument to `@RequestParam` for Spring Boot 3.2. See [56cebf9](https://github.com/jhipster/generator-jhipster/commit/56cebf91073faf9b855bbc8854794c148e4c9137). I found the answer [here](https://stackoverflow.com/a/25797744).

I wonder if it’s related to https://github.com/jhipster/generator-jhipster/pull/24295

> 
> /cc @mshima and @hide212131 since I know we had to do this for the generator-jhipster-native blueprint. It might not be necessary in the blueprint once this PR is merged.

Great. Maybe there are more customizations to move to main generator.

**Comment by mraible** [comment]
> I wonder if it’s related to https://github.com/jhipster/generator-jhipster/pull/24295

I believe it is.

**Comment by DanielFran** [comment]
@mraible we should be able to do a new release after November 28 based on the spring-cloud release
image

**Comment by mraible** [comment]
@DanielFran Sounds good to me! I think we should call it `8.1.0` since it will have Spring Boot 3.2 and Angular 17 included.

**Comment by mraible** [comment]
Some tests are failing with:

```
[0] Caused by: java.lang.NoSuchMethodError: 'long org.hibernate.engine.spi.SharedSessionContractImplementor.getTransactionStartTimestamp()'
[0] 	at org.infinispan.hibernate.cache.v60.impl.ReadOnlyEntityDataAccess.putFromLoad(ReadOnlyEntityDataAccess.java:30)
[0] 	at org.hibernate.sql.results.graph.entity.AbstractEntityInitializer.putInCache(AbstractEntityInitializer.java:1005)
[0] 	at org.hibernate.sql.results.graph.entity.AbstractEntityInitializer.updateCaches(AbstractEntityInitializer.java:912)
[0] 	at org.hibernate.sql.results.graph.entity.AbstractEntityInitializer.finishUpRow(AbstractEntityInitializer.java:1174)
[0] 	at org.hibernate.sql.results.internal.InitializersList.finishUpRow(InitializersList.java:65)
```

**Comment by mraible** [comment]
Here's an explanation the `name` argument for `@RequestParam`:

https://twitter.com/sam_brannen/status/1728019644802384343

**Comment by mraible** [comment]
It looks like the Swagger UI is broken. Navigating to Administration > API results in the following. 

Screenshot 2023-11-27 at 5 54 48 AM


Here's my `.yo-rc.json` you can use to re-create the project:

```json
{
  "generator-jhipster": {
    "applicationType": "monolith",
    "authenticationType": "jwt",
    "baseName": "blog",
    "buildTool": "maven",
    "cacheProvider": "ehcache",
    "clientFramework": "angular",
    "clientTestFrameworks": ["cypress"],
    "clientTheme": "none",
    "creationTimestamp": 1701088333195,
    "cypressAudit": true,
    "cypressCoverage": true,
    "databaseType": "sql",
    "devDatabaseType": "h2Disk",
    "devServerPort": 4200,
    "enableGradleEnterprise": null,
    "enableHibernateCache": true,
    "enableSwaggerCodegen": false,
    "enableTranslation": true,
    "entities": [],
    "gradleEnterpriseHost": null,
    "jhipsterVersion": "8.0.0",
    "languages": ["en", "es"],
    "messageBroker": false,
    "microfrontend": null,
    "microfrontends": [],
    "nativeLanguage": "en",
    "packageName": "org.jhipster.blog",
    "prodDatabaseType": "postgresql",
    "reactive": false,
    "searchEngine": false,
    "serverPort": null,
    "serverSideOptions": [],
    "serviceDiscoveryType": false,
    "testFrameworks": ["cypress"],
    "websocket": false,
    "withAdminUi": true
  }
}
```

**Comment by mraible** [comment]
I was able to solve the Swagger issue by deleting my browser's cache.

**Comment by mraible** [comment]
I ran into another issue that seems to be a blocker. In my tutorial, I say the following:

```
To add more security around blogs and entries, open `BlogResource.java` and find the `getAllBlogs()` method. Change the following line:

[source,java]
.src/main/java/org/jhipster/blog/web/rest/BlogResource.java
----
return blogRepository.findAll();
----

To:

[source,java]
----
return blogRepository.findByUserIsCurrentUser();
----

The `findByUserIsCurrentUser()` method is generated by JHipster in the `BlogRepository` class and allows limiting results by the current user.

[source,java]
.src/main/java/org/jhipster/blog/repository/BlogRepository.java
----
public interface BlogRepository extends JpaRepository<Blog, Long> {
    @Query("select blog from Blog blog where blog.user.login = ?#{principal.username}")
    List<Blog> findByUserIsCurrentUser();
}
----
```

After making this change, there's an "Internal server error." rendered in the UI and the logs have the following:

```
2023-11-27T07:09:05.963-07:00 ERROR 66062 --- [  XNIO-1 task-3] o.jhipster.blog.web.rest.BlogResource    : Exception in getAllBlogs() with cause = 'NULL' and exception = 'EL1008E: Property or field 'username' cannot be found on object of type 'org.springframework.security.oauth2.jwt.Jwt' - maybe not public or not valid?' 

org.springframework.expression.spel.SpelEvaluationException: EL1008E: Property or field 'username' cannot be found on object of type 'org.springframework.security.oauth2.jwt.Jwt' - maybe not public or not valid?
```

**Comment by mraible** [comment]
I've confirmed this happens with JHipster 8.0.0 (Spring Boot 3.1.5) too. It seems like the `DomainUserDetailsService` is not being used.

**Comment by mraible** [comment]
Spring Security's [docs](https://docs.spring.io/spring-security/reference/features/integrations/data.html#data-configuration) seem to indicate a `SecurityEvaluationContextExtension` bean is needed. I tried adding one to `SecurityConfiguration.java`, but it does not help:

```java
@Bean
public SecurityEvaluationContextExtension securityEvaluationContextExtension() {
    return new SecurityEvaluationContextExtension();
}
```

**Comment by mraible** [comment]
The current CI failures seem to be related to this again:

```
NoClassDefFoundError: Could not initialize class com.fasterxml.jackson.databind.ser.std.StdArraySerializers
```

I'm not sure how this was fixed last time. Maybe it's just a random failure?

I tried creating the `ng-webflux-psql-default` sample locally, and `npm run e2e` passes. 

**Update**: I used [this Stack Overflow Q&A](https://stackoverflow.com/questions/39419160/jackson-databind-classpath-issue) to do some further debugging. 

1. I expanded the JAR created with `mvn package` using `jar xvf jhipster-0.0.1-SNAPSHOT.jar`.
2. I cd'd into `BOOT-INF/lib` and ran `find . -type f -name '*.jar' -print0 |  xargs -0 -I '{}' sh -c 'jar tf {} | grep StdArraySerializers.class &&  echo {}'`

The result:

```
com/fasterxml/jackson/databind/ser/std/StdArraySerializers.class
./jackson-databind-2.15.3.jar
```

**Comment by mraible** [comment]
Proceeding with my jhipster8-demo tutorial. I found that the following refactoring to `PostResource#getPost()` no longer works:

```java
@GetMapping("/{id}")
public ResponseEntity<?> getPost(@PathVariable("id") Long id) {
    log.debug("REST request to get Post : {}", id);
    Optional<Post> post = postRepository.findOneWithEagerRelationships(id);
    if (post.isPresent() && post.get().getBlog() != null &&
        !post.get().getBlog().getUser().getLogin().equals(SecurityUtils.getCurrentUserLogin().orElse(""))) {
        return new ResponseEntity<>("Unauthorized", HttpStatus.UNAUTHORIZED);
    }
    return ResponseUtil.wrapOrNotFound(post);
}
```

The error is from Modernizer:

```
[ERROR] /Users/mraible/Downloads/blog/src/main/java/org/jhipster/blog/web/rest/PostResource.java:196: Prefer java.util.Optional.orElseThrow
```

The line it's referring to starts with is `if (post.isPresent() && post.get().getBlog() != null &&`. I'm not sure how to refactor this because I want to check if the blog exists, and if it does, then run the security check.

**Comment by mraible** [comment]
ChatGPT gave me an answer that seems to work:

```java
@GetMapping("/{id}")
public ResponseEntity<Post> getPost(@PathVariable("id") Long id) {
    log.debug("REST request to get Post : {}", id);
    Optional<Post> post = postRepository.findOneWithEagerRelationships(id);
    post
        .filter(p -> p.getBlog() != null && p.getBlog().getUser().getLogin().equals(SecurityUtils.getCurrentUserLogin().orElse("")))
        .orElseThrow(() -> new ResponseStatusException(HttpStatus.UNAUTHORIZED, "Unauthorized"));
    return ResponseUtil.wrapOrNotFound(post);
}
```

The problem is when I hit this method with an invalid ID, it throws a large stack trace and logs me out. If I use my original code, the console is much cleaner, but the logout still happens. 

With JHipster 7, the logout happens too. Maybe it's always been this way, and I never noticed?

**Comment by mraible** [comment]
I feel like I'm going down a rabbit hole that never ends. `principal.claims['sub']` doesn't work with tests:

```
{
  "type": "https://www.jhipster.tech/problem/problem-with-message",
  "title": "Internal Server Error",
  "status": 500,
  "detail": "EL1008E: Property or field 'claims' cannot be found on object of type 'org.springframework.security.core.userdetails.User' - maybe not public or not valid?",
  "instance": "/api/blogs",
  "message": "error.http.500",
  "path": "/api/blogs"
}
```

**Comment by mraible** [comment]
I reverted the change to use `principal.claims['sub']` since it's not related to this PR. I created a [new issue](https://github.com/jhipster/generator-jhipster/issues/24374) to track it.

**Comment by mraible** [comment]
@DanielFran The Spring Cloud 2023.0.0 release has moved to [December 5](https://github.com/spring-cloud/spring-cloud-release/milestone/132).

**Comment by mraible** [comment]
FWIW, I didn't mean to do a rebase and add all the commits from `main` individually. Doh!

File: generators/server/templates/pom.xml.ejs [hunk 1]
```diff
@@ -923,6 +923,28 @@
...
+        <pluginRepository>
+            <id>spring-milestones</id>
+            <name>Spring Milestones</name>
+            <url>https://repo.spring.io/milestone</url>
+            <snapshots>
+                <enabled>false</enabled>
+            </snapshots>
+        </pluginRepository>
+    </pluginRepositories>
+    <%_ } _%>
```
Comment (mshima) [inline generators/server/templates/pom.xml.ejs:947]:
```suggestion
```

Comment (mraible) [reply generators/server/templates/pom.xml.ejs:947]: How does it get added without it? Just curious for knowledge.

**Review by mshima — APPROVED** [review]

**Comment by mraible** [comment]
`[ms-ng-consul-oauth2-mongodb-caffeine](https://github.com/jhipster/generator-jhipster/actions/runs/7122578110/job/19395489593?pr=24098#logs)` seems to be a blocker. If it fails regularly, I'm fine with ignoring it.

**Comment by mraible** [comment]
Might be a [Kafka issue](https://github.com/jhipster/generator-jhipster/actions/runs/7122578110/job/19395489593#step:18:1) now?

**Comment by DanielFran** [comment]
> Might be a [Kafka issue](https://github.com/jhipster/generator-jhipster/actions/runs/7122578110/job/19395489593#step:18:1) now?

It seems, yes https://github.com/jhipster/generator-jhipster/actions/runs/7122578110/job/19395489593#step:18:712

image

**Comment by DanielFran** [comment]
Let's merge it and work separately in that issue

**Review by DanielFran — APPROVED** [review]

**Pull request merged**